File: src/examples/tutorial/ascent_intro/python/conduit_example4.py
###############################################################################
# Copyright (c) Lawrence Livermore National Security, LLC and other Ascent
# Project developers. See top-level LICENSE AND COPYRIGHT files for dates and
# other details. No copyright assignment is required to contribute to Ascent.
###############################################################################

import conduit
import numpy as np

#
# Conduit supports zero copy, allowing a Conduit Node to describe and
# point to externally allocated data.
# 
# set_external() is method used to zero copy data into a Node
#
n = conduit.Node()
a1 = np.zeros(10, dtype=np.int32)

a1[0] = 0
a1[1] = 1

for i in range(2,10):
   a1[i] = a1[i-2] + a1[i-1]


# create another array to demo difference 
# between set and set_external
a2 = np.zeros(10, dtype=np.int32) 

a2[0] = 0
a2[1] = 1

for i in range(2,10):
   a2[i] = a2[i-2] + a2[i-1]

n["fib_deep_copy"].set(a1);
n["fib_shallow_copy"].set_external(a2);

a1[-1] = -1
a2[-1] = -1

print(n.to_yaml())

File: src/examples/tutorial/ascent_intro/cpp/conduit_example4.cpp
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Copyright (c) Lawrence Livermore National Security, LLC and other Ascent
// Project developers. See top-level LICENSE AND COPYRIGHT files for dates and
// other details. No copyright assignment is required to contribute to Ascent.
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

//-----------------------------------------------------------------------------
///
/// file: conduit_example4.cpp
///
//-----------------------------------------------------------------------------

#include <iostream>
#include <vector>
#include "conduit.hpp"

using namespace conduit;
using std::vector;

int main()
{
    //
    // Conduit supports zero copy, allowing a Conduit Node to describe and
    // point to externally allocated data.
    //
    // set_external() is method used to zero copy data into a Node
    //

    Node n;
    vector<int> A1(10);
    A1[0] = 0;
    A1[1] = 1;
    for (int i = 2 ; i < 10 ; i++)
        A1[i] = A1[i-2]+A1[i-1];
    n["fib_deep_copy"].set(A1);

    // create another array to demo difference 
    // between set and set_external
    vector<int> A2(10);

    A2[0] = 0;  
    A2[1] = 1; 
    for (int i = 2 ; i < 10 ; i++)
    {
        A2[i] = A2[i-2]+A2[i-1];
    }

    n["fib_shallow_copy"].set_external(A2);
    A1[9] = -1;
    A2[9] = -1;

    std::cout << n.to_yaml() << std::endl;
}

